### ISSUE TITLE
Unable to use some environment variables

### ISSUE THREAD

**Issue by 0mmariano** [ansible/ansible#23489]
ISSUE TYPE: Bug report
COMPONENT NAME: ansible_env

ANSIBLE VERSION:
ansible 2.2.1.0
config file = /etc/ansible/ansible.cfg
configured module search path = ['/usr/share/my_modules/', '/etc/ansible/roles/glassfish/library']

CONFIGURATION:
defined values under [default]

OS: Debian Jessie

SUMMARY: Ansible is unable to interpolate an environment variable when it is called from a hosts file or a group_vars/all file.

STEPS TO REPRODUCE:

In /etc/profile.d/required_env_vars.sh --
DEPLOY_DIR=/myhome/deployments

In testPlaybook.yml
deploy_dir = "{{ ansible_env.DEPLOY_DIR }}/RT{{rt_number}}"
where rt_number is defined via vars_prompt

This fails because DEPLOY_DIR is not found in the output of the setup command for the host (looked in the ansible_env), but HOME is. So it looks like Ansible is unable to use environment variables defined in a profile.d script.

As a work-around, I added an [all:vars] section the inventories/poc/hosts file --
[all:vars]
deploy_dir = $HOME/deployments --> did not work
deploy_dir = /myhome/deployments --> worked
deploy_dir = $DEPLOY_DIR --> did not work
deploy_dir = "{{ ansible_env.DEPLOY_DIR }}" --> did not work

Since the second one worked (hardcoded value), I removed the [all:vars] section and used an inventories/poc/group_vars/all file --
deploy_dir = /myhome/deployments

This resulted in the error below --

ERROR! Unexpected Exception: dictionary update sequence element #0 has length 1; 2 is required
the full traceback was:

Traceback (most recent call last):
  File "/usr/bin/ansible-playbook", line 103, in <module>
    exit_code = cli.run()
  File "/usr/lib/python2.7/dist-packages/ansible/cli/playbook.py", line 132, in run
    inventory = Inventory(loader=loader, variable_manager=variable_manager, host_list=self.options.inventory)
  File "/usr/lib/python2.7/dist-packages/ansible/inventory/__init__.py", line 98, in __init__
    self.parse_inventory(host_list)
  File "/usr/lib/python2.7/dist-packages/ansible/inventory/__init__.py", line 165, in parse_inventory
    group.vars = combine_vars(group.vars, self.get_group_variables(group.name))
  File "/usr/lib/python2.7/dist-packages/ansible/inventory/__init__.py", line 555, in get_group_variables
    self._vars_per_group[groupname] = self._get_group_variables(groupname, vault_password=vault_password)
  File "/usr/lib/python2.7/dist-packages/ansible/inventory/__init__.py", line 573, in _get_group_variables
    vars = combine_vars(vars, self.get_group_vars(group))
  File "/usr/lib/python2.7/dist-packages/ansible/inventory/__init__.py", line 775, in get_group_vars
    return self._get_hostgroup_vars(host=None, group=group, new_pb_basedir=new_pb_basedir, return_results=return_results)
  File "/usr/lib/python2.7/dist-packages/ansible/inventory/__init__.py", line 839, in _get_hostgroup_vars
    host_results = self._variable_manager.add_group_vars_file(base_path, self._loader)
  File "/usr/lib/python2.7/dist-packages/ansible/vars/__init__.py", line 619, in add_group_vars_file
    data = self._load_inventory_file(path, loader)
  File "/usr/lib/python2.7/dist-packages/ansible/vars/__init__.py", line 577, in _load_inventory_file
    rval.update(data)
ValueError: dictionary update sequence element #0 has length 1; 2 is required
 
On my target server, I also tried to set up DEPLOY_DIR as an environment variable in various places --
1. /etc/environment -- which is not preferred because it can't do variable interpolation (only key=val pairs, no variables on the right hand side). The output of the env command showed the variables I defined as expected.
2. /etc/profile -- The output of the env command showed the variables defined also.
3. /etc/profile.d/required_env_vars.sh -- this is preferred because I'd like to keep /etc/profile the same across all servers, and just add custom environment variables via profile.d script. The output of the env command DID NOT show the variables defined, but doing "echo $variable" showed the expected value

None of these affected the content of ansible_env (I thought ansible_env would get the variables if the were returned by the "env" command?).

So it looks like Ansible does not include all the environment variables defined in its "environment" (ansible_env).

**UPDATE 1**

Okay. My bad. I changed the content of group_vars/all to use a colon instead of an equal and it worked.

deploy_dir : /home/b013000915/deployments

Got confused because in some cases (e.g. facts.d content, some modules), I had to use an equal sign.

But the other issue (not getting some of the environment variables defined) still persists.

**UPDATE 2**

I also tried using lookup to get the value of the following variables on the target node but it did not work for custom variables (defined via /etc/profile)

In /etc/profile
DEPLOY_DIR=/myhome/deployments

In testPlaybook.yml
- name: check using lookup env
    debug:
      msg: "deploy_dir is {{ lookup('env', 'DEPLOY_DIR') }}, home is {{ lookup('env', 'HOME') }}"

Output was
ok: [IP address] => {
    "msg": "deploy_dir is , home is /myhome"
}

**Comment by abadger** [comment]
* lookups operate on the controller and thus will use environment variables from the controller.  Only modules operate on the remote side. (Second note in the documentation of lookups: http://docs.ansible.com/ansible/playbooks_lookups.html )

* Different shells source different startup files so I can't say for certain which startup file you'll need to change.  I can say that I've just tested under bash and zsh on Ubuntu, Fedora, and CentOS (I don't have a Debian VM otherwise I would have tested there for you as well) and using /bin/bash will make use of ~/.bashrc while /bin/zsh will use ~/.zshenv  In general, which startup files the shells are sourcing when ansible uses them via ssh is out of scope for us but I figured I'd mention my findings since they fresh in my memory.

Hope that helps!

**Comment by 0mmariano** [comment]
@abadger - Thanks for your help!

I tried the same changes on the controller and got the same results: 

1. /etc/environment -- The lookup found the variable defined
2. /etc/profile -- The lookup also found the variable defined
3. /etc/profile.d/required_env_vars.sh -- The lookup did not find the variable. One of the things that /etc/profile does when it is read is call each *.sh file in /etc/profile.d

Out of curiosity, I added "export" to the entry in /etc/profile.d/required_env_vars.sh --
export DEPLOY_DIR=/myhome/deployments

This time, the Ansible lookup found it. "echo $<variable>" returns the value in the command line whether you have "export" or not in the profile.d/*.sh but Ansible was only able to find variables explicitly set with the "export" command.
